pallets/popi/src/lib.rs

```diff
@@ -262,19 +259,79 @@ pub mod pallet {
 				.ok_or(Error::<T>::UserExperienceDoesNotExist)
 		}
 
-		fn update_user_experience(
+		pub fn update_user_experience(
 			user: T::AccountId,
-			experience: UserExperience<T>,
+			exp_type: ExperienceType,
+			mut experience: UserExperience<T>,
 		) -> DispatchResult {
-			unimplemented!()
+			// check if the user has experience
+			if !ExperienceStorage::<T>::contains_key((&user, &exp_type)) {
+				return Err(Error::<T>::UserExperienceDoesNotExist.into());
+			}
+
+			experience.level = Self::calculate_exp_level(experience.experience, experience.level)?;
+			let next_exp = Self::calc_exp_of_level(experience.level + 1)?;
+			experience.exp_to_next_lvl = Self::remaining_exp(experience.experience, next_exp)?;
+
+			// otherwise, update the user's experience
+			ExperienceStorage::<T>::set((user, &exp_type), Some(experience));
+			Ok(())
 		}
 
-		/// Usees our Config types to calculate the amount of experience required to level up
-		fn calculate_exp_to_next_level(
-			experience: UserExperience<T>,
+		/// Uses our Config types to calculate the amount of experience required to level up
+		/// BaseExperience * DifficultyMultiplier ^ (LevelDifficulty * (level - 1))
+		pub fn calculate_exp_level(
+			// Current experience of the user.
+			experience: u128,
+			// Current level of the user.
 			level: u32,
-		) -> DispatchResult {
-			unimplemented!()
+		) -> Result<u32, DispatchError> {
+			let mut new_level = level;
+			let mut count: usize = 0;
+
+			loop {
```

Comment:
Could you try instead an eager algorithm?
I'm thinking, as the experience of the user is a result of positive interactions, why not on each of those interactions, once registered as positive, why not eventually calculating the experience in this very moment?

Reply:
I'm not completely sure what you mean. I considered hard-coding the levels to prevent the running of this computation, but I obviously opted not to do that in the end. Is that what you're suggesting? We just hard-code the levels instead of having the on-chain calculations?

Reply:
if you describe me what is this loop doing, perhaps I try to explain my idea. Thanks :)

Reply:
Sure! Let's assume you are a new user (level 0) and work on a large project where you earn 1500 experience all at once. Without a loop, you would be leveled up to level 1, even though level 1 only requires 100 xp. The loop allows a calculation of multiple levels at once. So it will iterate through until it knows you can't level up again. 

That's where the `count` variable comes in. There should NEVER be a time where someone levels up that many times at once. I wanted that to be a final security measure. If someone finds a way to cheat the system, it at least will not allow an infinite loop to run.

Reply:
this proposal from you is really interesting. It is suggesting new challenges:

1. 
when you say "Let's assume you are a new user (level 0) and work on a large project where you earn 1500 experience all at once."

this is considering user X worked on a big project. Big project means for example worked on task1, task2, task3....taskN right?

why not increasing the XP on each ticket? 

2.
we could think on other kind of limits: maximum one level increase per time? maximum one positive interaction every x hours. etc. I need to talk to some friends in the gaming industry, I remember some discussions about balancing that could come useful in this moment.

Reply:
I see a loop that repeat 10 times maximum right? And on each loop it is calling a method calc_exp_of_level. So theoretically for one task a person may level up up to 10 levels for one single task.
This is what initially looked messy to me. And for this reason I'm trying to suggest something a bit more linear.
Now, when you say "each level requires an exponential amount of experience, so this shouldn't be a problem". What is limiting a single task from giving an exponential amount of experience to one person?

Reply:
I haven't built anything to implement restrictions on experience per task. That's another issue we can take a look at. Let's set up a time to talk through this via chat. I think there's a disconnect here, maybe on my end. I want to make sure I'm following completely. What do you say? @virgiliolino @DeluxeRaph